File: scala/src/main/scala/clase3/Functor.scala
package clase3

import scala.io.StdIn

object Functor extends App {

  def readAnInt(message:String) = {
    val str = StdIn.readLine(message)

    try {
      Some(str.toInt)
    } catch {
      case ex: NumberFormatException => None
    }
  }

  val maybeInt = readAnInt("Enter an Int that you'd like to double:\n")
  val maybeDoubled = maybeInt.map(_ * 2)

  println(maybeDoubled)

  maybeDoubled match {
    case Some(x) => println("here is your double: " + x)
    case None => println("you are a mean person")
  }
}

File: haskell/src/clase3/Functor.hs
module Functor where

import Text.Read

readInt :: String -> Maybe Int
readInt line = readMaybe line :: Maybe Int

mapDouble = fmap (\int -> int * 2)

desc (Just x) = "your double is " ++ (show x)
desc (Nothing) = "you are a mean person"

main = do
  print "Enter a number to double"
  line <- getLine
  print (
    let maybeDouble = mapDouble (readInt line)
    in (show maybeDouble) ++ " -> " ++ (desc maybeDouble))
  print "lists are functors too"
  print ((2*) <$> [1, 2, 3, 4])
